subsys/bootloader/bl_monotonic_counters/bl_monotonic_counters.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bl_monotonic_counters.h>
+#include <string.h>
+#include <errno.h>
+#include <nrf.h>
+#include <assert.h>
+#include <nrfx_nvmc.h>
+
+#ifdef CONFIG_SECURE_BOOT_STORAGE
+#include <bl_storage.h>
+#else
+#include <pm_config.h>
+#endif
+
+#ifdef CONFIG_MCUBOOT_HW_DOWNGRADE_PREVENTION_COUNTER_SLOTS
+BUILD_ASSERT(CONFIG_MCUBOOT_HW_DOWNGRADE_PREVENTION_COUNTER_SLOTS % 2 == 0,
+			 "CONFIG_MCUBOOT_HW_DOWNGRADE_PREVENTION_COUNTER_SLOTS was not an even number");
+#endif
+
+#ifdef CONFIG_SB_NUM_VER_COUNTER_SLOTS
+BUILD_ASSERT(CONFIG_SB_NUM_VER_COUNTER_SLOTS % 2 == 0,
+			 "CONFIG_SB_NUM_VER_COUNTER_SLOTS was not an even number");
+#endif
+
+
+/** This library implements monotonic counters where each time the counter
+ *  is increased, a new slot is written.
+ *  This way, the counter can be updated without erase. This is, among other things,
+ *  necessary so the counter can be used in the OTP section of the UICR
+ *  (available on e.g. nRF91 and nRF53).
+ */
+struct monotonic_counter {
+	uint16_t description; /* Counter description. What the counter is used for. See BL_MONOTONIC_COUNTERS_DESC_*. */
+	uint16_t num_counter_slots; /* Number of entries in 'counter_slots' list. */
+	uint16_t counter_slots[1];
+};
+
+/**  This struct is placed after the bl_storage_data struct if present. It has unknown length since
+ *  'counters' is repeated. Note that each entry in counters also has unknown
+ *  length, and each entry can have different length from the others, so the
+ *  entries beyond the first cannot be accessed via array indices.
+ */
+struct counter_collection {
+	uint16_t type; /* Must be "monotonic counter". */
+	uint16_t num_counters; /* Number of entries in 'counters' list. */
+	struct monotonic_counter counters[1];
+};
+
+/** Get the counter_collection data structure in the provision data. */
+static const struct counter_collection *get_counter_collection(void)
+{
+#if defined(CONFIG_SECURE_BOOT_STORAGE)
+	const struct counter_collection *collection = (struct counter_collection *)
+		&BL_STORAGE->key_data[num_public_keys_read()];
+#else
+	const struct counter_collection *collection = (struct counter_collection *)PM_PROVISION_ADDRESS;
+#endif
+
```

Comment:
This is overly complicated.

We should encode the position of the counters into PM to simplify this.